aws/request/timeout_read_closer.go

```diff
@@ -0,0 +1,65 @@
+package request
+
+import (
+	"io"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+var timeoutErr = awserr.New(
+	awserr.CodeConnectionTimeout,
+	"read on body has reached the timeout limit",
+	nil,
+)
+
+type readResult struct {
+	n   int
+	err error
+}
+
+// timeoutReadCloser will handle body reads that take too long.
+// We will return a ErrReadTimeout error if a timeout occurs.
+type timeoutReadCloser struct {
+	reader   io.ReadCloser
+	duration time.Duration
+}
+
+// Read will spin off a goroutine to call the reader's Read method. In addition,
+// we use the time.AfterFunc to return an timeout error once a duration of time
+// has occurred.
+func (r *timeoutReadCloser) Read(b []byte) (int, error) {
+	timer := time.NewTimer(r.duration)
+	c := make(chan readResult, 1)
+
+	go func() {
+		n, err := r.reader.Read(b)
+		// cancel any AfterFunc pending calls.
+		timer.Stop()
+		c <- readResult{n: n, err: err}
+	}()
+
+	select {
+	case data := <-c:
+		return data.n, data.err
+	case <-timer.C:
+		return 0, timeoutErr
+	}
+}
+
+func (r *timeoutReadCloser) Close() error {
+	return r.reader.Close()
+}
+
+// WithTimeoutReadCloser is a request option that will wrap the body in a timeout read closer.
+// This will allow for per packet timeouts.
+//
+//     svc.PutObjectWithContext(ctx, params, request.WithTimeoutReadCloser(30 * time.Second)
+func WithTimeoutReadCloser(duration time.Duration) Option {
+	return func(req *Request) {
+		req.HTTPResponse.Body = &timeoutReadCloser{
```

Comment:
I don't think this will do the right thing.  Request Options are added at the beginning of a request lifecycle. This will cause a nil pointer panic because HTTPResponse isn't a valid value until after the `Send` handlers are executed.

I suggest this be changed to:

```go
func WithResponseReadTimeout(duration time.Duration) Option {
    return func(r *Request) {
        r.Handlers.Send.PushBack(func(req *Request) {
            req.HTTPResponse.Body = &timeoutReadCloser{
                reader:    req.HTTPResponse.Body,
                duration: duration,
            }
        })
    }
}
```